lib/plugin.js

```diff
@@ -1,108 +1,113 @@
-const cds = require("@sap/cds/lib");
-const LOG = cds.log("attachments");
-const { extname } = require("path");
-const DEBUG = LOG._debug ? LOG.debug : undefined;
-const attachmentIDRegex = /\/\w+\(.*ID=([0-9a-fA-F-]{36})/;
-
-cds.on("loaded", function unfoldModel(csn) {
-  if (!("Attachments" in csn.definitions)) return;
-  const csnCopy = structuredClone(csn)
-  cds.linked(csnCopy).forall("Composition", (comp) => {
+const cds = require("@sap/cds/lib")
+const LOG = cds.log("attachments")
+const { extname } = require("path")
+const DEBUG = LOG._debug ? LOG.debug : undefined
+const attachmentIDRegex = /\/\w+\(.*ID=([0-9a-fA-F-]{36})/
+
+cds.on("compile.to.edmx", unfoldModel)
+// cds.on("loaded", unfoldModel) //> would still be required for older releases of @sap/cds
+
+function unfoldModel (csn) {
+  const meta = csn.meta ??= {}
+  if (!("Attachments" in csn.definitions)) return
+  if (meta._enhanced_for_attachments) return
+  // const csnCopy = structuredClone(csn) // REVISIT: Why did we add this cloning?
```

Comment:
This was done because when we convert `csn` to `LinkedCSN` (next line) the `@source` property is deleted from the definitions of kind `service` in the `csn`. This was leading to deployment failures in production with the below error when files were present inside a folder in `srv` and not directly in `srv`.

`Error: Failed loading service implementation from '/home/vcap/app/srv/<file-name>.js'` 

See `cap/issues#16004` (cannot link it here) for more details.

Reply:
Thanks @sidakphull → will fix the root cause instead of having to go such workarounds.

Reply:
Hi @sidakphull, as you already pointed out, the issues might be due to `@source` being by default. 

As I was not able to reproduce the erroneous behaviour, though, may I ask you to do so with this line changed like that, in you local `node_modules/@sap/cds/lib/srv/factory.js`:

```diff
-- const _source = (d) => d['@source'] || (d['@source'] = d.$location && d.$location.file.replace(/\\/g, '/') || '.')
++ const _source = (d) => d['@source'] || d.$location?.file || '.'
```

Reply:
not able to reproduce locally since I don't have the sample application from the issue anymore and it seems to be deleted. Will try again and get back to you regarding this.